src/dbnode/persist/fs/commitlog/reader.go

```diff
@@ -207,9 +206,64 @@ func (r *reader) readLogEntry() error {
 	return nil
 }
 
+func (r *reader) namespaceIDReused(id []byte) ident.ID {
+	var namespaceID ident.ID
+	for _, ns := range r.namespacesRead {
+		if bytes.Equal(ns.namespaceIDBytes, id) {
+			namespaceID = ns.namespaceIDRef
+			break
+		}
+	}
+	if namespaceID == nil {
+		// Take a copy and keep around to reuse later.
+		namespaceBytes := append(make([]byte, 0, len(id)), id...)
+		namespaceID = ident.BytesID(namespaceBytes)
+		r.namespacesRead = append(r.namespacesRead, namespaceRead{
+			namespaceIDBytes: namespaceBytes,
+			namespaceIDRef:   namespaceID,
+		})
+	}
+	return namespaceID
+}
+
 func (r *reader) seriesMetadataForEntry(
 	entry schema.LogEntry,
-) (seriesMetadata, error) {
+) (ts.Series, error) {
+	if r.opts.returnMetadataAsRef {
+		// NB(r): This is a fast path for callers where nothing
+		// is cached locally in terms of metadata lookup and the
+		// caller is returned just references to all the bytes in
+		// the backing commit log file the first and only time
+		// we encounter the series metadata, and then the refs are
+		// invalid on the next call to metadata.
+		if len(entry.Metadata) == 0 {
+			// Valid, nothing to return here and caller will already
+			// have processed metadata for this entry (based on the
+			// FileReadID and the SeriesUniqueIndex returned).
+			return ts.Series{}, nil
```

Comment:
Might be being a bit paranoid here, but would it make sense to add something like a `ShouldBeLoaded` param to `ts.Series` to better understand why the fields related to these are unset?